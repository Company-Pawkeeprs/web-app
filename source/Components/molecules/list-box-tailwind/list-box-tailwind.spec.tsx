import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import { Formik } from 'formik';
import ListBoxTailwind, { ListBoxTailwindProps } from './list-box-tailwind';


const wrapper = (children: React.ReactNode | JSX.Element) => (
    <Formik initialValues={{}} onSubmit={() => { }}>
        {children}
    </Formik>
)

const sampleItems = [
    { value: 1, name: 'Item 1' },
    { value: 2, name: 'Item 2' },
    { value: 3, name: 'Item 3' },
]

const defaultProps: ListBoxTailwindProps<{}> = {
    items: sampleItems,
    name: 'test-listbox',
}

describe('List Box Tailwind Snapshot (Unit)', () => {
    it('should match snapshot', () => {
        const { container } = render(wrapper(<ListBoxTailwind items={[]} name='pet' />))

        expect(container).toMatchSnapshot()
    })

    it('should match snapshot with options', async () => {
        const { container } = render(wrapper(<ListBoxTailwind items={sampleItems} name='pet' />))

        const button = screen.getByTestId('list-box-tailwind')

        userEvent.click(button)

        await new Promise((resolve) => setTimeout(resolve, 500))

        expect(container).toMatchSnapshot()

    })

    it('should match snapshot with a option selected', async () => {
        const { container } = render(wrapper(<ListBoxTailwind option={{ name: 'Selected on Option', value: 1 }} {...defaultProps} />))

        await new Promise((resolve) => setTimeout(resolve, 500))

        expect(container).toMatchSnapshot()
    })

    it('should match snapshot with a option selected', async () => {
        const item = { name: 'Selected on Option', value: 1 }
        const newItem = { name: 'Selected on Option 3', value: 2 }
        const { rerender, container } = render(wrapper(<ListBoxTailwind option={item} {...defaultProps} />))

        await new Promise((resolve) => setTimeout(resolve, 500))
        rerender(wrapper(<ListBoxTailwind option={newItem} {...defaultProps} />))

        expect(container).toMatchSnapshot()
    })
})

describe('List Box Tailwind (Unit)', () => {



    it('should render label', async () => {
        render(wrapper(<ListBoxTailwind items={[]} name='pet' label='Pet' />))

        expect(screen.getByText('Pet')).toBeInTheDocument()
    })

    it('should disables the ListBoxTailwind component when disabled prop is true', async () => {
        render(wrapper(<ListBoxTailwind {...defaultProps} disabled />))

        const button = screen.getByRole('button')
        expect(button).toBeDisabled()
    })

    it('should renders the ListBoxTailwind component with a placeholder', async () => {
        const placeholder = 'Select an item'
        render(wrapper(<ListBoxTailwind placeholder={placeholder} {...defaultProps} />))

        expect(screen.getByText(placeholder)).toBeInTheDocument()
    })


    it('should renders the ListBoxTailwind component with a option selected', async () => {
        const option = { name: 'Selected', value: 1 }
        render(wrapper(<ListBoxTailwind option={option} {...defaultProps} />))

        expect(screen.getByText(option.name)).toBeInTheDocument()
    })

    it('renders the ListBoxTailwind component with the given items', async () => {
        render(wrapper(<ListBoxTailwind {...defaultProps} />))

        const button = screen.getByTestId('list-box-tailwind')

        userEvent.click(button)

        await waitFor(() => {
            sampleItems.forEach((item) => {
                expect(screen.getByText(item.name)).toBeInTheDocument()
            })
        }, { timeout: 500 })

    })

    it('selects an item from the ListBoxTailwind component', async () => {
        render(wrapper(<ListBoxTailwind {...defaultProps} />))
        const button = screen.getByTestId('list-box-tailwind')
        userEvent.click(button)

        await waitFor(() => {
            const optionToSelect = screen.getByTestId('list-box-tailwind-options')
            expect(optionToSelect).toBeInTheDocument()
        }, { timeout: 200 })
    })

    it('should select an item from the ListBoxTailwind component', async () => {
        render(wrapper(<ListBoxTailwind {...defaultProps} />))

        const button = screen.getByTestId('list-box-tailwind')
        userEvent.click(button)

        await new Promise((resolve) => setTimeout(resolve, 500))

        sampleItems.forEach(async (item, index) => {
            const optionToSelect = screen.getByTestId(`list-box-tailwind-options-${index}`)
            userEvent.click(optionToSelect)

            await new Promise((resolve) => setTimeout(resolve, 500))

            expect(screen.getByText(item.name)).toBeInTheDocument()
        })

    })

    it('should select an item from the ListBoxTailwind component rerender option', async () => {
        const item = { name: 'Selected on Option', value: 1 }
        const { rerender } = render(wrapper(<ListBoxTailwind option={item} {...defaultProps} />))
        expect(screen.getByText(item.name)).toBeInTheDocument()

        const newItem = { name: 'Selected on Option 2', value: 2 }

        rerender(wrapper(<ListBoxTailwind option={newItem} {...defaultProps} />))

        expect(screen.getByText(newItem.name)).toBeInTheDocument()

    })

    it('should renders the ListBoxTailwind component with a placeholder default if not option', async () => {
        const placeholder = 'Select an item'
        const item = {}

        render(wrapper(<ListBoxTailwind placeholder={placeholder} option={item as any} {...defaultProps} />))

        expect(screen.getByText(placeholder)).toBeInTheDocument()
    })


})

describe('List Box Tailwind Hooks (Unit)', () => {
    it('should call onChange when an item is selected', async () => {
        const onChange = vi.fn()
        render(wrapper(<ListBoxTailwind {...defaultProps} onChangeOption={onChange} />))

        const button = screen.getByTestId('list-box-tailwind')
        userEvent.click(button)

        await new Promise((resolve) => setTimeout(resolve, 500))

        const optionToSelect = screen.getByTestId('list-box-tailwind-options-0')
        userEvent.click(optionToSelect)

        await new Promise((resolve) => setTimeout(resolve, 500))
        expect(onChange).toHaveBeenCalled()
    })

})