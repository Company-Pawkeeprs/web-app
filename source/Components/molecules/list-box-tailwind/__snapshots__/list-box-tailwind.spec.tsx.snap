// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List Box Tailwind Snapshot (Unit) > should match snapshot 1`] = `
<div>
  <div>
    <div
      class="relative mt-1"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="
                                    relative w-full cursor-default 
                                    rounded-lg bg-white dark:!bg-dark
                                    py-2 pl-3 pr-10 text-left 
                                    shadow-md focus:outline-none 
                                    focus-visible:border-indigo-500 
                                    focus-visible:ring-2 
                                    focus-visible:ring-white 
                                    focus-visible:ring-opacity-75 
                                    focus-visible:ring-offset-2 
                                    focus-visible:ring-offset-orange-300 
                                    sm:text-sm
                                    disabled:opacity-50
                                    disabled:cursor-not-allowed
                                    disabled:dark:opacity-70"
        data-headlessui-state=""
        data-testid="list-box-tailwind"
        id="pet-list-box-tailwind"
        placeholder="Selecione uma opção"
        type="button"
      >
        <span
          class="block truncate"
        >
          Selecione uma opção
        </span>
        <span
          class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 text-gray-400"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`List Box Tailwind Snapshot (Unit) > should match snapshot with a option selected 1`] = `
<div>
  <div>
    <div
      class="relative mt-1"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="
                                    relative w-full cursor-default 
                                    rounded-lg bg-white dark:!bg-dark
                                    py-2 pl-3 pr-10 text-left 
                                    shadow-md focus:outline-none 
                                    focus-visible:border-indigo-500 
                                    focus-visible:ring-2 
                                    focus-visible:ring-white 
                                    focus-visible:ring-opacity-75 
                                    focus-visible:ring-offset-2 
                                    focus-visible:ring-offset-orange-300 
                                    sm:text-sm
                                    disabled:opacity-50
                                    disabled:cursor-not-allowed
                                    disabled:dark:opacity-70"
        data-headlessui-state=""
        data-testid="list-box-tailwind"
        id="test-listbox-list-box-tailwind"
        placeholder="Selecione uma opção"
        type="button"
      >
        <span
          class="block truncate"
        >
          Selected on Option
        </span>
        <span
          class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 text-gray-400"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`List Box Tailwind Snapshot (Unit) > should match snapshot with a option selected 2`] = `
<div>
  <div>
    <div
      class="relative mt-1"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="
                                    relative w-full cursor-default 
                                    rounded-lg bg-white dark:!bg-dark
                                    py-2 pl-3 pr-10 text-left 
                                    shadow-md focus:outline-none 
                                    focus-visible:border-indigo-500 
                                    focus-visible:ring-2 
                                    focus-visible:ring-white 
                                    focus-visible:ring-opacity-75 
                                    focus-visible:ring-offset-2 
                                    focus-visible:ring-offset-orange-300 
                                    sm:text-sm
                                    disabled:opacity-50
                                    disabled:cursor-not-allowed
                                    disabled:dark:opacity-70"
        data-headlessui-state=""
        data-testid="list-box-tailwind"
        id="test-listbox-list-box-tailwind"
        placeholder="Selecione uma opção"
        type="button"
      >
        <span
          class="block truncate"
        >
          Selected on Option 3
        </span>
        <span
          class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 text-gray-400"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`List Box Tailwind Snapshot (Unit) > should match snapshot with options 1`] = `
<div>
  <div>
    <div
      class="relative mt-1"
    >
      <button
        aria-controls="headlessui-listbox-options-:r2:"
        aria-expanded="true"
        aria-haspopup="listbox"
        class="
                                    relative w-full cursor-default 
                                    rounded-lg bg-white dark:!bg-dark
                                    py-2 pl-3 pr-10 text-left 
                                    shadow-md focus:outline-none 
                                    focus-visible:border-indigo-500 
                                    focus-visible:ring-2 
                                    focus-visible:ring-white 
                                    focus-visible:ring-opacity-75 
                                    focus-visible:ring-offset-2 
                                    focus-visible:ring-offset-orange-300 
                                    sm:text-sm
                                    disabled:opacity-50
                                    disabled:cursor-not-allowed
                                    disabled:dark:opacity-70"
        data-headlessui-state="open"
        data-testid="list-box-tailwind"
        id="pet-list-box-tailwind"
        placeholder="Selecione uma opção"
        type="button"
      >
        <span
          class="block truncate"
        >
          Selecione uma opção
        </span>
        <span
          class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 text-gray-400"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
      <ul
        aria-labelledby="pet-list-box-tailwind"
        aria-orientation="vertical"
        class="absolute mt-1 max-h-60 w-full overflow-auto rounded-md bg-white dark:!bg-gray-700 py-1 text-base shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none sm:text-sm z-5u0"
        data-headlessui-state="open"
        data-testid="list-box-tailwind-options"
        id="headlessui-listbox-options-:r2:"
        role="listbox"
        tabindex="0"
      >
        <li
          aria-selected="false"
          class="relative cursor-default select-none py-2 pl-10 pr-4 text-gray-900 dark:text-gray-200"
          data-headlessui-state=""
          data-testid="list-box-tailwind-options-0"
          id="headlessui-listbox-option-:r3:"
          role="option"
          tabindex="-1"
        >
          <span
            class="block truncate font-normal"
          >
            Item 1
          </span>
        </li>
        <li
          aria-selected="false"
          class="relative cursor-default select-none py-2 pl-10 pr-4 text-gray-900 dark:text-gray-200"
          data-headlessui-state=""
          data-testid="list-box-tailwind-options-1"
          id="headlessui-listbox-option-:r4:"
          role="option"
          tabindex="-1"
        >
          <span
            class="block truncate font-normal"
          >
            Item 2
          </span>
        </li>
        <li
          aria-selected="false"
          class="relative cursor-default select-none py-2 pl-10 pr-4 text-gray-900 dark:text-gray-200"
          data-headlessui-state=""
          data-testid="list-box-tailwind-options-2"
          id="headlessui-listbox-option-:r5:"
          role="option"
          tabindex="-1"
        >
          <span
            class="block truncate font-normal"
          >
            Item 3
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
